### ISSUE TITLE
Inversion mode per block [Enchancement]

### ISSUE THREAD

**Issue by leodluz** [mlaily/NegativeScreen#17]
Make inversion perblocks on screen, so that darker regions still darker, and white regions are inverted.

**Comment by mlaily** [comment]
Hi!

Could you please elaborate a bit what you have in mind?

**Comment by leodluz** [comment]
![image](https://user-images.githubusercontent.com/22651765/57412331-becbc780-71c6-11e9-8808-424f89975984.png)
Reverse only in areas that are clear, as in the above example (photoshop). The regions could be delimited by blocks of 100x100 pixels, so the program would make an analysis of the average brightness of that block. If the light average of the block is greater than 0.5 make an inversion. For example blocks that have brightness 0.7, with the inversion would be 0.3. Blocks that are already dark will remain dark. That way we would have a totally nightmode desktop.

**Comment by mlaily** [comment]
Ok I see. I thought about doing this kind of thing, but never actually tested it...

Unfortunately I see a few problems that might kill the idea:
- The block size is very important, because it's unlikely the borders of a block will align with the borders of a bright window like in your example. So ideally the blocks should be the smallest possible. But if you choose a block size small enough, there will be no bright region remaining on screen. I'm pretty sure an entirely dark screen is not very useful and not what we want. There might be some kind of compromise to do here, but the fundamental problem is still there.
- The performance of creating hundreds or thousands of small blocks will probably not be great. It might not be the bottleneck though, read on.
- Each screen region corresponding to a block needs to be averaged independently. Almost all of the screen has to be analyzed, and it takes time. Doing it once should still be okay, but with this mode, we can't only do it once. We have to analyze the screen continually to avoid screen regions staying bright or dark when things move. At this point the performance would likely not be acceptable anymore...

**Comment by leodluz** [comment]
What if you do not analyze each block completely? a 100x100 block contains 10 thousand pixels, and if you analyze only 25 pixels (using a for i+=25)? would reduce loads and give an average of satisfactory accuracy.

![image](https://user-images.githubusercontent.com/22651765/57454283-d47cd380-723e-11e9-90ac-883d8a91b66f.png)

Or Perhaps this is the case where only gpu shader would solve. But applying a shader over windows may not be simple.

Thank you for answering me, God bless you.

**Comment by Comawhite54** [comment]
Is it possible to define an area that will never be negatived ? I would like that my taskbar always stays in the standard color.

**Comment by mlaily** [comment]
> 
> 
> Is it possible to define an area that will never be negatived ? I would like that my taskbar always stays in the standard color.

No, sorry. It's only possible to create surfaces to apply color transformation, not the opposite. Excluding a rectangular zone would require creating 8 zones around the hole...

**Comment by Comawhite54** [comment]
If I understand, it is possible to create a rectangle of which area I want to apply the transform ? Then it is also ok for me (I will draw a rectangle that cover my whole screen except the taskbar). How can I do that?

**Comment by mlaily** [comment]
The magnifier api has two modes: full screen, and normal. ([called 'control' in the documentation](https://docs.microsoft.com/en-us/previous-versions/windows/desktop/magapi/magapi-intro#types-of-magnifiers)).

The full screen mode is easier to use, more efficient/performant, and handles some edge cases better (e.g. it's always above everything else, so pressing windows+tab or alt+tab does not create flicker like the magnifier control mode does)

In control mode, you have to define the magnifier control, initialize it, then refresh it manually. This is how the [legacy version](https://github.com/mlaily/NegativeScreen/tree/legacy_version) of NegativeScreen used to work. You can see how a magnifier control is defined for each screen here: https://github.com/mlaily/NegativeScreen/blob/legacy_version/NegativeScreen/NegativeOverlay.cs

If you want to use this mode, I suggest you start from the `legacy_version` branch (or the `brightness_detection` one) instead of  the main branch (the newest version does not support the control mode anymore).

**Comment by kr99** [comment]
I have migraine triggers from bright screens. So this is an accessibility thing for me.  On a highly collaborative team, I have coworkers constantly sharing their screen on Zoom, switching between bright screens (chrome/jira/github) and dark ones (terminals/IDEs/etc) (depending on their preferences).  With 3 monitors at home, I can't switch _everything_

I think you should simplify this feature as follows with one or more of these options:

- Allow a _user_ to toggle negation _per active window_.  The user has control, and NegativeScreen does not need to do much in the way of smarts--it's left to the user.
- Allow a _user_ to toggle negation _per monitor/display_.
- If NegativeScreen wants to get smart, it could:
   - Do a once-only analysis of window regions/blocks _when a user presses a hotkey_.  
   - It could start with larger blocks, 
   - It can subdivide blocks on some "good enough" condition" When it finds a block with some condition (greater than 10% deviation?), it could subdivide the blocks and re-analyze.  
   - Subdivide _on request with a hotkey_, it _retains the settings for that window_, and it _zeroes in on black/white areas progressively_ .   
   - If the user resizes/moves a window, they could press the hotkey to undo/redo the analysis/etc.

I strongly suggest **starting with the minimal, easiest feature**, then adding to the features progressively.   Today, I have money I can pay for the most basic version of this feature.

As far as having 8 blocks surrounding something... I am guessing you can come up with an abstraction in your code to make that easy to manage?

If you'd like to pair-code on this, I'd be happy to offer a couple hours of help here and there (nodejs & java experience here).  Maybe that would be a payment in kind at least.

**Comment by mlaily** [comment]
Thanks for your feedback. It's always useful!

**Comment by kr99** [comment]
Bumping this--I really want this feature.

I would do the simplest implementation possible (minimum viable feature) first.  Then see how it's used and enhance if/when needed.

To simplify it even more than previously noted, you could allow negating different monitors separately with a new hotkey binding.  (This makes it so you don't have to figure out active windows, their geometry, work with special windows, window movement/resizes, etc).  Just figure out the cursor location (or something) when win-shift-m (next to n) is pressed, work out the geometry of the current monitor, and flip it.  If someone docks/undocks, I guess that's an event to watch for?  Not sure how that works.

**Comment by DavidHJong** [comment]
Is there any GUI for toggling per-monitor/app color inversion? If so, can anyone show me the steps to invert color of one monitor/app? 
(I don't understand much tech, so I can't grasp the answer from your previous discussions.)
